ament_cmake_cppcheck/cmake/ament_cmake_cppcheck_lint_hook.cmake

```diff
@@ -24,5 +24,31 @@ file(GLOB_RECURSE _source_files FOLLOW_SYMLINKS
 )
 if(_source_files)
   message(STATUS "Added test 'cppcheck' to perform static code analysis on C / C++ code")
-  ament_cppcheck()
+
+  # Get include paths for added targets
+  set(_all_include_dirs "")
+  # BUILDSYSTEM_TARGETS only supported in CMake >= 3.7
+  if(NOT CMAKE_VERSION VERSION_LESS "3.7.0")
+    get_directory_property(_build_targets DIRECTORY ${CMAKE_SOURCE_DIR} BUILDSYSTEM_TARGETS)
+    foreach(_target ${_build_targets})
+      get_property(_include_dirs
+        TARGET ${_target}
+        PROPERTY INCLUDE_DIRECTORIES
+      )
+
+      # Only append include directories that are from the package being tested
+      # This accomplishes two things:
+      #     1. Reduces execution time (less include directories to search)
+      #     2. cppcheck will not check for errors in external packages
+      foreach(_include_dir ${_include_dirs})
+        string(REGEX MATCH "^${CMAKE_SOURCE_DIR}.*" _is_match ${_include_dir})
+        if(_is_match)
+          list(APPEND _all_include_dirs ${_include_dir})
+        endif()
+      endforeach()
+    endforeach()
+    list(REMOVE_DUPLICATES _all_include_dirs)
```

Comment:
If I remember correctly I looked into this function in the past and at that time it didn't guarantee stability of the result. Since it is important that the order of the include dirs is maintained I would suggest to use [list_append_unique](https://github.com/ament/ament_cmake/blob/2d65d41e047e203cd0531df9b0987d8aaea7d6ca/ament_cmake_core/cmake/core/list_append_unique.cmake#L28) instead of this line and `list(APPEND ...)` above.

Reply:
I don't think the order of the include directories matters in this case, does it?
I can make the suggested change nonetheless.

Reply:
The order of the include directories will determine which header file will be used if the same file exists in multiple locations.

Reply:
Done